example/echo-op/echo_op.cpp

```diff
@@ -217,18 +201,84 @@ operator()(beast::error_code ec, std::size_t bytes_transferred)
             break;
     }
 
-    // Invoke the final handler. If we wanted to pass any arguments
-    // which come from our state, they would have to be moved to the
-    // stack first, since the `handler_ptr` guarantees that the state
-    // is destroyed before the handler is invoked.
+    // Invoke the final handler. The implementation of `handler_ptr`
+    // will deallocate the storage for the state before the handler
+    // is invoked. This is necessary to provide the
+    // destroy-before-invocation guarantee on handler memory
+    // customizations.
+    //
+    // If we wanted to pass any arguments to the handler which come
+    // from the `state`, they would have to be moved to the stack
+    // first or else undefined behavior results.
     //
     p_.invoke(ec);
     return;
 }
 
 //]
 
-//[example_core_echo_op_2
+//[example_core_echo_op_6
+
+// Handler hook forwarding. These free functions invoke the hooks
+// associated with the final completion handler. In effect, they
+// make the Asio implementation treat our composed operation the
+// same way it would treat the final completion handler for the
+// purpose of memory allocation and invocation.
+//
+// Our implementation just passes through the call to the hook
+// associated with the final handler.
```

Comment:
Consider moving the comments from `asio_handler_is_continuation()` here.

> Our implementation just passes ~through~ the call **through** to the hook associated with the final handler. Note that we structure the calls to permit argument dependent lookup to take effect (`using boost::asio::asio_handler_*;`).
>
> Also, always use std::addressof to pass the pointer to the handler, otherwise an unwanted overload of operator& may be called instead.